File: cpp/739. Daily Temperatures.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// Monotonic decreasing deque, Time: O(n), Space: O(n)
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        int n = T.size();
        deque<int> d; // We store idx in deque, the values are decreasing
        vi ans(n);
        for(int i = n-1; i >=0; i--) {
            while(d.size() && T[d.back()] <= T[i]) d.pop_back();
            if(d.size()) ans[i] = d.back() - i;
            else ans[i] = 0;
            d.push_back(i);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi T, ans;

    T = {73, 74, 75, 71, 69, 72, 76, 73};
    ans = solution.dailyTemperatures(T);
    assert(ans==vi({1, 1, 4, 2, 1, 1, 0, 0}));
}

File: python/739-Daily Temperatures.py
from typing import List

class Solution:
    def dailyTemperatures(self, T: List[int]) -> List[int]:
        # Monotonic Decreasing Deque, Time: O(n), Space: O(n)
        from collections import deque;
        d = deque() # We store idx in deque, the values are decreasing
        n = len(T)
        ans = [0]*n
        for i in reversed(range(n)):
            while(d and T[d[-1]] <= T[i]): d.pop()
            if(d): ans[i] = d[-1] - i
            else: ans[i] = 0
            d.append(i)
        return ans

T = [73, 74, 75, 71, 69, 72, 76, 73]
ans = Solution().dailyTemperatures(T)
assert ans == [1, 1, 4, 2, 1, 1, 0, 0]